parallel unit test runner for Windows and Unix

klausspanderen opened the pull request:
Running QuantLib’s test suite on a recent computer takes around 10min. The situation will improve a lot if the test runner utilises more than one core. Unfortunately multi-threading won’t work because the boost unit test framework is not thread safe. 

This boost unit test runner spawns several child processes and distributes the work/results via boost interprocess.

The runner has been successfully tested under Windows and Linux but Peter has reported issues under OSX.

pcaspers commented:
just a note, if somebody wants to try the runner on OSX (I have 10.11.2), you probably have to outcomment this line 

https://github.com/klausspanderen/QuantLib/blob/master/acinclude.m4#L264

because there doesn't seem to be a librt (and not needed for boost::interprocess) on OSX.

klausspanderen commented:
thanks Peter for the hint. I've changed the script to first try to use boost::interprocess without librt and only if this fails try to link with librt. This should work on OSX

@Luigi, I've also changed the code for the boost_thread_lib to solve issue #35: Detect alternate naming for Boost libraries

lballabio commented:
Cool, thanks.

lballabio commented:
Cool, thanks.

Closes #35

klausspanderen commented:
Hi Luigi
 
could you do me a favour? Peter has reported some issues with this patch on OSX
but I don't have access to a Mac. We are not sure whether the problems are due
to Peter's local installation or a general problem of the patch (maybe boost
interprocess). Could you give it a try on your Mac?
 
regards
klaus

> Luigi Ballabio <notifications@github.com> hat am 19. Januar 2016 um 11:59
> geschrieben:
> 
> 
>  Cool, thanks.
> 
>  —
>  Reply to this email directly or view it on GitHub
> https://github.com/lballabio/QuantLib/pull/40#issuecomment-172817924 .
> 
>   
>

lballabio commented:
Ok

pcaspers commented:
btw, boost does have own unit-tests, I assume, do either of you know how to execute them? This way I could check the interprocess installation (it would be interesting anyhow)

lballabio commented:
@pcaspers : no, I never looked into running them.

klausspanderen commented:
@pcaspers: I've never tested boost libraries. Maybe a last thing to check on OSX: W.r.t. your last issue report: The issue would be explainable if the mutex does not work as expected. Could you please remove line 148, 149

       static named_mutex mutex(open_or_create, "namesLogMutexName");
       scoped_lock<named_mutex> lock(mutex);

pcaspers commented:
no, unfortunately that does not change anything, I get the same seg fault

lballabio commented:
Not just Peter's problem&mdash;it segfaults on my Mac as well.

lballabio commented:
I'll go ahead and merge it though, so it's available on Windows and Linux. We can keep investigating and fix it on OS X later.

lballabio commented:
@klausspanderen : at this time there are errors on master (and they are logged when the tests run) but `make check` exits successfully on my Ubuntu box. Maybe there's a problem while communicating the results back to the runner?

lballabio commented:
Or maybe it's just the return code returned by `main` that doesn't reflect the result?

klausspanderen commented:
I have to admit that I've never used make check before. You are right, the parallel test runner does not return the overall result. I'll send a PR with the necessary change.
